prov/tcp/src/xnet_cm.c

```diff
@@ -51,18 +51,21 @@ struct xnet_cm_entry {
  * and is the first data transferred over the socket.
  */
 static int
-xnet_recv_cm_msg(SOCKET sock, struct xnet_cm_msg *msg, uint8_t exp_msg)
+xnet_recv_cm_msg(SOCKET sock, struct xnet_cm_msg *msg, uint8_t exp_msg,
+		 bool skip_hdr)
 {
 	size_t len;
 	ssize_t ret;
 
-	len = sizeof(msg->hdr);
-	ret = ofi_recv_socket(sock, &msg->hdr, len, 0);
-	if ((size_t) ret != len) {
-		FI_WARN(&xnet_prov, FI_LOG_EP_CTRL,
-			"Failed to read cm header\n");
-		ret = ofi_sockerr() ? -ofi_sockerr() : -FI_EIO;
-		goto err;
+	if (!skip_hdr) {
+		len = sizeof(msg->hdr);
+		ret = ofi_recv_socket(sock, &msg->hdr, len, 0);
+		if ((size_t) ret != len) {
+			FI_WARN(&xnet_prov, FI_LOG_EP_CTRL,
+				"Failed to read cm header\n");
+			ret = ofi_sockerr() ? -ofi_sockerr() : -FI_EIO;
+			goto err;
+		}
```

Comment:
I think it would it be better to separate the call into 2 functions, with the first receiving the header.  The io_uring path does this.  This function would then become more of a 'handle_cm_msg' than recv call then.

Reply:
Thanks. The new revision implements your suggestion.